filebeat/input/kafka/input.go

```diff
@@ -0,0 +1,285 @@
+				select {
+				case <-input.context.Done:
+					return
+				default:
+				}
+			}
+		}()
+	})
+}
+
+// Wait shuts down the Input by cancelling the internal context.
+func (input *kafkaInput) Wait() {
+	input.Stop()
+	// Wait for the consumer group to shut down
+	input.consumerGroup.Close()
+}
+
+// Stop shuts down the Input by cancelling the internal context.
+func (input *kafkaInput) Stop() {
+	close(input.context.Done)
+}
+
+func arrayForKafkaHeaders(headers []*sarama.RecordHeader) []interface{} {
+	array := []interface{}{}
+	for _, header := range headers {
+		array = append(array, common.MapStr{
+			"key":   header.Key,
+			"value": header.Value,
+		})
+	}
+	return array
+}
+
+// A barebones implementation of context.Context wrapped around the done
+// channels that are more common in the beats codebase. This could be added
+// as a utility in a shared part of the code, but right now it's a special
+// case for sarama which requires a context.Context, so it's private until
+// there's at least one other use case.
+type channelCtx <-chan struct{}
+
+func doneChannelContext(ch <-chan struct{}) context.Context {
+	return channelCtx(ch)
+}
+
+func (c channelCtx) Deadline() (deadline time.Time, ok bool) { return }
+func (c channelCtx) Done() <-chan struct{} {
+	return (<-chan struct{})(c)
+}
+func (c channelCtx) Err() error {
+	select {
+	case <-c:
+		return context.Canceled
+	default:
+		return nil
+	}
+}
+func (c channelCtx) Value(key interface{}) interface{} { return nil }
+
+type groupHandler struct {
+	input *kafkaInput
+}
+
+func (h groupHandler) createEvent(
+	sess sarama.ConsumerGroupSession,
+	claim sarama.ConsumerGroupClaim,
+	message *sarama.ConsumerMessage,
+) beat.Event {
+	event := beat.Event{
+		Timestamp: time.Now(),
+	}
+	eventFields := common.MapStr{
+		"message": string(message.Value),
+	}
+	kafkaMetadata := common.MapStr{
+		"topic":     claim.Topic(),
+		"partition": claim.Partition(),
+		"offset":    message.Offset,
+		"key":       message.Key,
+	}
+	version, versionOk := h.input.config.Version.Get()
+	if versionOk && version.IsAtLeast(sarama.V0_10_0_0) {
+		event.Timestamp = message.Timestamp
+		if !message.BlockTimestamp.IsZero() {
+			kafkaMetadata["block_timestamp"] = message.BlockTimestamp
+		}
+	}
+	if versionOk && version.IsAtLeast(sarama.V0_11_0_0) {
+		kafkaMetadata["headers"] = arrayForKafkaHeaders(message.Headers)
+	}
+	eventFields["kafka"] = kafkaMetadata
+	event.Fields = eventFields
+	return event
+}
+
+func (h groupHandler) Setup(session sarama.ConsumerGroupSession) error {
+	h.input.sessionState.setSession(session)
+	return nil
+}
+
+func (h groupHandler) Cleanup(_ sarama.ConsumerGroupSession) error {
+	h.input.sessionState.setSession(nil)
+	return nil
+}
+
+func (h groupHandler) ConsumeClaim(sess sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	for msg := range claim.Messages() {
+		event := h.createEvent(sess, claim, msg)
+		h.input.outlet.OnEvent(event)
+		sess.MarkMessage(msg, "")
```

Comment:
I'm confused, do we still need to call 'MarkMessage' here?

Reply:
nope, this just didn't get updated >_< thanks!